### DISCUSSION TITLE
Building datoviz on Windows.

### DISCUSSION THREAD

**Discussion by ron-adam** [datoviz/datoviz#73] [General]
_Originally posted by @ron-adam in https://github.com/datoviz/datoviz/issues/72#issuecomment-2560193003_

Moving this thread from issues due to the conversation not being related to that specific issue. 

I've been able to build Datoviz on Ubuntu, and am able to build the main branch on windows.   But still can't build using the dev branch.   [rossant](https://github.com/rossant) is working on this, so it should be resolved soon.

My specific problem was due to not using the correct version of imgui files.  Version 1.91.0 was needed to build for windows.   It may be good to have a release page that documents more specifically what version of external files are being used in 'main'.  And to specify if that will change in the 'dev' branch.   Adding a version tag for the 'data' may also be good when dev is merged with main.

**Comment by rossant** [comment]
Could you explain what you mean by "imgui files"? I think imgui should be integrated as a git submodule so there shouldn't be any need for external imgui-related files/dependencies. If there is, I think it's a bug, not a feature!

**Reply by ron-adam** [reply]
It may be I'm not using git quite right.  On windows, when I use "git clone", I get an empty directory where the symbolic links are.   The symbolic link points to a specific commit ```imgui @ 6864a7f````.

My first attempt, I downloaded a zip of the imgui repository and used it to supply the files in that directory.   The build process worked, but the resulting build did not work correctly.   The zip file I used supplied the most current version of imgui, and not the state it was in at commit 6864a7f.   I then checked earlier version tags of imgui until I found the one that seemed to match the files and dates.  "Version 1.91.0"   Then instead of using the zip file, I went to the empty imgui directory and did a git clone of imgui with that version tag.  That resulted in a successful and working build on windows for the main Datoviz branch.

This could be a bug if the intended behavior is for the build process to work with the most current version of imgui.

Or possibly a documentation issue if a specific version is required, or alternative build instructions are needed.

Or an error on my part if, there is a commonly known way to get windows to resolve symbolic links as it does on Linux.

Either way, I think a note for what version of dependencies used at each release is good practice.   What I've seen before is that projects tend to pick a major release version (with long term support) for dependent software.   It looks like imgui isn't organized in that way, but you could select the most recent version at each datoviz release.  (That results in a build where all of Datoviz's tests pass.)  Documenting this can be useful in determining if bugs are in Datoviz, or the dependent software.  It may also be useful information for projects that depend on Datoviz.  

Looking at the dev branch, it currently uses the same link as main, which may result in using an older version of imgui than you intend.  Newer features may not be available, or may not work correctly, at that specific point. 

Apologies if I'm missing something obvious.

I currently use Vispy for a real time display window.   The general description of Datoviz matches some of my needs,  but isn't developed enough just yet.   So I'm hoping I can offer some development support.   I can possibly help with Python, Testing, and Documentation.

**Reply by rossant** [reply]
I should update the Windows build instructions to specify that recursive cloning is required:

`git clone https://github.com/datoviz/datoviz.git --recursive`

A standard clone is not going to work (crazy that it's not the default behavior in git).

This is explained in the Linux and macOS instructions. On Windows, I use the GitHub desktop GUI application which does that automatically.

**Reply by ron-adam** [reply]
Thank you,  that worked.

Looking in the imgui doc/changlog.txt file now shows:
```
-----------------------------------------------------------------------
 VERSION 1.91.1 WIP (In Progress)
-----------------------------------------------------------------------
```
I used 1.91.0 which was close enough to work, but the --recursive flag is a better solution.

Version 1.91.1 WIP,  is also the current the version used in the dev branch.  So it will be the version used for the next Datoviz main branch if it's not changed.

I would avoid the WIP between release states if possible, but it may not matter that much if the differences are small.

Do you have a policy for when to use newer releases?   
https://github.com/ocornut/imgui/releases/

**Reply by rossant** [reply]
Good point, I would actually stick to stable releases of imgui at all times in both main and dev. If that's not the case now, we should fix it.

**Reply by ron-adam** [reply]
The --recursive flag solved part of my issues, however I was still having trouble building a successful windows datoviz (main branch) build.

I figured there was still a version conflict of some sort.  The demo and examples windows would run and work but disappear right after resizing.

I managed to get a working build a few minutes ago by using an earlier version of the vulkan sdk.   After uninstalling version Vulkan SDK 1.3.296, and installing Vulkan SDK 1.3.290,  I was able to do a successful "just build", "just test" and "just demo" with no failures from a windows cmd line. (Not a bash shell).   I believe the Ubuntu vulkan files are an even earlier version.

Is there a specific Vulkan SDK version that I should be using?

Doing a "just wheel" received an import dataviz error during the checkstructs call.   I temporarily commented it out to build a wheel.  And then was able to run the python examples after installing it.   The checksize recipe attempts to import datoviz, but datoviz isn't available until after it's installed.   Looks like this is already fixed in the dev branch. 

(Note: In the dev branch  "tools/struct_sizes.c" file is missing a few commas.  You probably already found and corrected that in your local files.)

I am still getting validation errors that don't occur in the Ubuntu version, this too may be vulkan version mismatch issue.  (?)
```
(.venv) D:\datoviz_development\datoviz\examples>just demo
./build/datoviz demo
14:50:46.104 T0 W D:\datoviz_development\datoviz\src\vkutils.h:0183: validation layer: Validation Warning: [ BestPractices-specialuse-extension ] | MessageID = 0x675dc32e | vkCreateInstance():  Attempting to enable extension VK_EXT_debug_utils, but this extension is intended to support use by applications when debugging and it is strongly recommended that it be otherwise avoided.
14:50:46.156 T0 W D:\datoviz_development\datoviz\src\vkutils.h:0183: validation layer: Validation Warning: [ WARNING-DEBUG-PRINTF ] | MessageID = 0x76589099 | vkCreateDevice():  Internal Warning: Forcing VkPhysicalDeviceFeatures::fragmentStoresAndAtomics to VK_TRUE
14:50:46.156 T0 W D:\datoviz_development\datoviz\src\vkutils.h:0183: validation layer: Validation Warning: [ WARNING-DEBUG-PRINTF ] | MessageID = 0x76589099 | vkCreateDevice():  Internal Warning: Forcing VkPhysicalDeviceFeatures::vertexPipelineStoresAndAtomics to VK_TRUE
14:50:46.157 T0 W D:\datoviz_development\datoviz\src\vkutils.h:0183: validation layer: Validation Warning: [ WARNING-DEBUG-PRINTF ] | MessageID = 0x76589099 | vkCreateDevice():  Internal Warning: Adding a VkPhysicalDeviceTimelineSemaphoreFeatures to pNext with timelineSemaphore set to VK_TRUE
creating volume texture (1372 bytes)
14:50:46.414 T0 W D:\datoviz_development\datoviz\src\vkutils.h:0183: validation layer: Validation Warning: [ BestPractices-ImageBarrierAccessLayout ] Object 0: handle = 0x73a850000000004d, type = VK_OBJECT_TYPE_IMAGE; | MessageID = 0xf35d019f | vkCmdPipelineBarrier(): pImageMemoryBarriers[0] image is VkImage 0x73a850000000004d[] and accessMask is VK_ACCESS_2_MEMORY_READ_BIT, but for layout VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL expected accessMask are VK_ACCESS_2_INPUT_ATTACHMENT_READ_BIT|VK_ACCESS_2_SHADER_READ_BIT|VK_ACCESS_2_COLOR_ATTACHMENT_READ_BIT|VK_ACCESS_2_SHADER_SAMPLED_READ_BIT|VK_ACCESS_2_SHADER_STORAGE_READ_BIT|VK_ACCESS_2_SHADER_BINDING_TABLE_READ_BIT_KHR.

[and more...]
```

The present build instructions for Windows seem to be a how to build a linux wheel on a windows machine.  I did the above in a standard windows command prompt (Not Bash).  The result was a "datoviz-0.2.2.dev0-py3-none-win_amd64.whl".

I hope these notes are useful.

**Comment by rossant** [comment]
Thank you for these details. I'll have a closer look next week.

We should better document which Vulkan SDK is used for the builds. I think it may be good practice to always use the latest version if possible (?).

The build instructions for Windows are primarily designed for creating a Windows build. However, I found it easier to use the mingw64 compiler instead of MSVC, as it aligns more closely with what I’m accustomed to on Ubuntu. Could you clarify which parts of the instructions seem to relate to building a Linux wheel on a Windows machine?

**Reply by ron-adam** [reply]
Yes, I agree about the Vulkan SDK.   However, I think past versions of datoviz, should use the versions they where developed with.

Would it be possible to add a 'just versions" recipe that reports the versions of build tools and external components being used?  That could make it easier to compare builds to each other, and identify what works, and what doesn't. 

I'm not an expert on this, but have watched the python development release cycle when I was active there.  The very wide usage required keeping older versions working for longer periods with a slower development process overall.  I'm not sure that applies for datoviz yet. 

On Linux systems, the installers do a good job of handling dependencies and avoiding combinations that are problematic.  It's not always the most current versions, but often ones that have been used for a while.

I don't think mingw64 is the problem.  It should work.

This is interesting.  
```
# From windows command line:
D:\>where just
C:\mingw64\bin\just.exe
C:\Users\ra\AppData\Local\Microsoft\WinGet\Packages\Casey.Just_Microsoft.Winget.Source_8wekyb3d8bbwe\just.exe

# From the bash shell:
D:\>bash
ron@DESKTOP-6N81HOL:/mnt/d$ which just
/home/ron/bin/just

# If  just.exe is used from bash:
ron@DESKTOP-6N81HOL:/mnt/d$ which just.exe
/mnt/c/mingw64/bin/just.exe
```
It appears that the "just" executable version determines which build recipe is used. (?)

Part of the confusion for me was in determining what the intended behavior was suppose to be.   By using linux bash shell, it suggests it is cross compiling, but then the windows versions of mingw64 is installed.  But when I did "just build", I got the Ubuntu version instead due to having a linux 'just' command present.

I think I should have caught that sooner.   I'm going to start again, and see if that helps.

**Reply by rossant** [reply]
I've always been confused by the various terminals on Windows. I would love to come up with better and less ambiguous build instructions.

**Reply by ron-adam** [reply]
The standard windows command prompt is suitable for this, and it's always available.

The instructions below work for me on windows.
After installing the requirements, the rest is nearly identical to the Ubuntu build instructions.
````
## Windows

Requirements:

* [Git for Windows](https://git-scm.com/download/win).
* [WinLibs](https://winlibs.com/): download and install the latest gcc UCRT version with POSIX threads..
* [just](https://github.com/casey/just/releases).
* [LunarG Vulkan SDK for Windows](https://vulkan.lunarg.com/sdk/home#windows).
* [vcpkg](https://vcpkg.io/en/). The `VCPKG_ROOT` environment variable should be set and should be in the `PATH`.
* [Python](https://www.python.org/downloads/windows/)

Instructions:

1. Install the above requirements if they are missing.
2. Copy the decompressed `just.exe` into `C:\mingw64\bin` (which should have been created by WinLibs).
3. Open windows command prompt at the location where you want to create the build.

```
# Clone the Datoviz repo.
git clone https://github.com/datoviz/datoviz.git --recursive
cd datoviz

# Build Python environment.
python -m venv .venv
.venv\scripts\activate
pip install -r requirements-dev.txt

# NOTE: this call will fail, but the build will succeed the second time.
# Fix welcome (see https://github.com/Chlumsky/msdf-atlas-gen/issues/98)
just build

# That one should succeed.
just build

# Try a demo.
just demo

# Compile and run a C example.
just example scatter

# Run the demo from Python.
python -c "import datoviz; datoviz.demo()"
```
````

Adding instruction on how to build and install a wheel may be good.  Include the --no-cache-dir flag on the pip command to avoid reusing a previous installed wheel.

On the dev branch, I believe I'm at the same point you are.  Ctypes won't load the libdatoviz.dll file, preventing the module from being imported.  I'll look at that further over the next few days if I have time.

Hope this helps.

**Reply by rossant** [reply]
Is the standard command prompt cmd.exe or powershell?

Do git, python, and just work directly in it after installing the requirements, without additional steps?

**Comment by ron-adam** [comment]
Yes, the standard cmd.exe, or right click start and select 'command prompt'.   (They are the same.)

I haven't needed additional steps.

You could add more specific instructions for setting the environment variable "VCPKG_ROOT" and adding it to the path.   Or wait until someone indicates it's an issue needing to be solved.

After the "just" command starts, it runs shell scripts with ```#!user/bin/env sh```.  On my computer that is the git bash shell.  That will be the same weather cmd.exe or powershell is used after Git for windows is installed.
```
C:\Users\ra>where sh
C:\Program Files\Git\usr\bin\sh.exe
```

Best to stick with simpler cmd.exe instructions for most users.  If they want to use a different shell, they will know how.

**Reply by rossant** [reply]
Okay, thanks. Please feel free to open a PR for improving the build instructions, by the way!

**Reply by ron-adam** [reply]
I will.   It appears I had some of the dependencies already installed.  I'm going to verify the installation steps on a clean machine.  And make sure the instructions for setting up the paths and environment variables are correct.   (As I get time, so it may take a few days.)

**Comment by rossant** [comment]
It seems that disabling OPENMP support fixes the problem on Windows (setting `DATOVIZ_WITH_OPENMP` to `OFF`). I don't know why, but that's a temporary workaround (OpenMP is not so much used in Datoviz at the moment).

**Reply by ron-adam** [reply]
That didn't work for me on the dev branch.   I presume you mean the setting near the top of the CMakeLists.txt file?   Or is there a different place that needs to be changed?

I'm currently using PyCharm as an ide.  It works well with github and mingw64 (and has many github capabilities build in.)  Creating local branches at any commit point is very easy for testing.

I think using the git-bash shell is a good build instruction choice.   The cmd.com terminal requires cygpath.exe from CygWin.  I had installed that and then forgot it.  It's an additional step that can be avoided if the git-bash shell is used.

The git-bash shell can work with python virtual environments, but requires either "source " or ". " preceding the activate command.
```bash
py -m venv .venv
source .venv/scripts/activate
pip install -r reqirements-dev.txt
```
A freshly installed python interpreter on windows doesn't add python to the path, but instead adds the 'py' launcher.
After the virtual environment is activated, then 'python' can be used as that is added to the local environment.

As for the updated build instructions,  I can do that.  It looks like the WSL2 addition possibly was meant to be so the automatic build tests can work from it and got added to the windows build instructions.  (I'm not sure about this. I'm still learning where and why everything is as they are.)

Context for developer builds may matter.
  1.  They are fine with the prebuild pip install.
  2.  Want a more recent build, but only for either Windows, Linux or Macos.
  3.  Or they are a datoviz contributor, and want or need to test multiple versions.
 
In the last case, A note in the contribution instructions will be enough.  The BUILD.md instructions only need to address the first 2 cases.

And separately:
The checkstucts justfile recipe needs a small change.  it currently imports datoviz, but that import is either not built yet, or uses the module installed from a previous build.  The fix is to insert the current working directory to the start of sys.path.  It will then import the datoviz module from the local datovize module directory before looking elsewhere, and that will be the current build which just completed.  I can do a pull request for this one.  (Easy fix.)

**Reply by ron-adam** [reply]
It may be the work around didn't work for me due to me not having other changes you have.
```bash
ra-dev@DESKTOP MINGW64 ~/PycharmProjects/datoviz (test-build)
$ ./build/datoviz demo
C:/Users/ra-dev/PycharmProjects/datoviz/build/datoviz.exe: error while loading shared libraries: libdatoviz.dll: cannot open shared object file: No such file or directory
```

**Reply by ron-adam** [reply]
I think the error I'm getting is a setting that was changed on my end.
Your work around may work.  Will let you know when I get my glitch here sorted out. (Probably tomorrow.)

**Reply by rossant** [reply]
I had similar issues, might be due to the cmake cache, make sure to completely clear the build folder and delete any intermediate auto-generated cmake files.

**Reply by ron-adam** [reply]
Ok.  I found the problem I was having.

For some reason (yet to be determined), the dlls in ```build/vcpkg_installed/x64-windows/debug/bin``` where not being found or copied to the build directory.  The easy way to determine which ones is, in windows explorer, double click on the datoviz.exe file in the build directory.  It will pop up a window with an error saying which one it can not find.  After copying the required dll's to the build directory, it would worked with no rebuild needed.  And with DATOVIZ_WITH_OPENMP set to OFF, it will also then build a wheel.

The strange part is why it was working earlier for me, but then stopped.  It may indicate the paths are not fully relative and the parent directory name or location was changed in some way.

**Comment by rossant** [comment]
Can this be closed @ron-adam? Are there any documentation improvements needed here?
